src/pk.c

```diff
@@ -6959,793 +7110,1199 @@ WOLFSSL_BIGNUM* wolfSSL_DH_8192_prime(WOLFSSL_BIGNUM* bn)
+ * @return  0 on failure.
+ * @return  1 on success.
+ */
+int wolfSSL_DH_set_length(WOLFSSL_DH *dh, long len)
 {
-    WOLFSSL_BIO* fbio = NULL;
-    WOLFSSL_DH* dh = NULL;
+    int ret = 1;
 
-    if (fp == NULL) {
-        WOLFSSL_MSG("DH parameter file cannot be NULL");
-        return NULL;
+    WOLFSSL_ENTER("wolfSSL_DH_set_length");
+
+    /* Validate parameter. */
+    if (dh == NULL) {
+        WOLFSSL_ERROR_MSG("Bad function arguments");
+        ret = 0;
+    }
+    else {
+        /* Store length. */
+        dh->length = (int)len;
     }
 
-    fbio = wolfSSL_BIO_new(wolfSSL_BIO_s_file());
-    if (fbio == NULL) {
-        WOLFSSL_MSG("Unable to create file BIO to process DH PEM");
-        return NULL;
+    return ret;
+}
+#endif /* OPENSSL_ALL || (v1.1.0 or later) */
+#endif
+
+/* Get the public and private keys requested.
+ *
+ * @param [in]  dh         DH key to get keys from.
+ * @param [out] pub_key    Pointer to return public key in. May be NULL.
+ * @param [out] priv_key   Pointer to return private key in. May be NULL.
+ */
+void wolfSSL_DH_get0_key(const WOLFSSL_DH *dh, const WOLFSSL_BIGNUM **pub_key,
+    const WOLFSSL_BIGNUM **priv_key)
+{
+    WOLFSSL_ENTER("wolfSSL_DH_get0_key");
+
+    /* Get only when valid DH passed in. */
+    if (dh != NULL) {
+        /* Return public key if available and available. */
+        if ((pub_key != NULL) && (dh->pub_key != NULL)) {
+            *pub_key = dh->pub_key;
+        }
+        /* Return private key if required and available. */
+        if ((priv_key != NULL) && (dh->priv_key != NULL)) {
+            *priv_key = dh->priv_key;
+        }
     }
+}
 
-    if (wolfSSL_BIO_set_fp(fbio, fp, BIO_NOCLOSE) != 1) {
-        wolfSSL_BIO_free(fbio);
-        WOLFSSL_MSG("wolfSSL_BIO_set_fp error");
-        return NULL;
+/* Set the public and/or private key.
+ *
+ * @param [in, out] dh        DH key to have keys set into.
+ * @param [in]      pub_key   Public key to set. May be NULL.
+ * @param [in]      priv_key  Private key to set. May be NULL.
+ * @return  0 on failure.
+ * @return  1 on success.
+ */
+int wolfSSL_DH_set0_key(WOLFSSL_DH *dh, WOLFSSL_BIGNUM *pub_key,
+    WOLFSSL_BIGNUM *priv_key)
+{
+    int ret = 1;
+
+    WOLFSSL_ENTER("wolfSSL_DH_set0_key");
+
+    /* Validate parameters. */
+    if (dh == NULL) {
+        ret = 0;
     }
 
-    /* wolfSSL_PEM_read_bio_DHparams() sanitizes x, cb, u args */
-    dh = wolfSSL_PEM_read_bio_DHparams(fbio, x, cb, u);
-    wolfSSL_BIO_free(fbio);
-    return dh;
+    /* Replace public key when one passed in. */
+    if ((ret == 1) && (pub_key != NULL)) {
+        wolfSSL_BN_free(dh->pub_key);
+        dh->pub_key = pub_key;
+    }
+
+    /* Replace private key when one passed in. */
+    if ((ret == 1) && (priv_key != NULL)) {
+        wolfSSL_BN_free(dh->priv_key);
+        dh->priv_key = priv_key;
+    }
+
+    /* Update internal key if parameters set. */
+    if ((ret == 1) && ((dh->p != NULL) && (dh->g != NULL))) {
```

Comment:
Should this be checking priv_key and pub_key rather than p and g? Also shouldn't it be using SetIndividualInternal for each individually, in case one is NULL?

Reply:
Changed to call SetIndividualInternal() when each is set.